spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,395 @@
+      Integer width = readAndValidateWidth(input);
+
+      Byte toTruncate = !input.isNullAt(1) ? input.getByte(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateSmallInt extends TruncateBase<Short> {
+    // magic method used in codegen
+    public static short invoke(int width, short value) {
+      return TruncateUtil.truncateShort(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.ShortType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.ShortType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](smallint)";
+    }
+
+    @Override
+    public Short produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      Short toTruncate = !input.isNullAt(1) ? input.getShort(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateInt extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int width, int value) {
+      return TruncateUtil.truncateInt(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.IntegerType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.IntegerType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](int)";
+    }
+
+    @Override
+    public Integer produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      Integer toTruncate = !input.isNullAt(1) ? input.getInt(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateBigInt extends TruncateBase<Long> {
+    // magic function for usage with codegen
+    public static long invoke(int width, long value) {
+      return TruncateUtil.truncateLong(width, value);
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.LongType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.LongType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](bigint)";
+    }
+
+    @Override
+    public Long produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      Long toTruncate = !input.isNullAt(1) ? input.getLong(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateString extends TruncateBase<UTF8String> {
+    // magic function for usage with codegen
+    // TODO - This can likely be made more efficient, but using existing definition for now.
```

Comment:
How would you make this more efficient?

Reply:
I think we could use the built-in `substring` method or `substringSQL`, which both seem to take into account the logic of `numChars()` method.

`substring`  specifically mentions that it works by code-point: https://github.com/apache/spark/blob/d77bc7027675b9823bcee96589eb0dddee3cfaa0/common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java#L264-L294

Reply:
Javadoc for `UTF8String::substring`

```
  /**
   * Returns a substring of this.
   * @param start the position of first code point
   * @param until the position after last code point, exclusive.
   */
```

Reply:
Tried it with `substring` and it passes all of the tests, including utf-8 unicode character aware truncation tests.

Updated to use that method and reverting the other previous changes.